### ISSUE TITLE
AvroModels interact strangely with Spark

### ISSUE THREAD

**Issue by nathan-oplus** [marcosschroh/dataclasses-avroschema#250]
This library is really wonderful but I'm getting some strange behavior that I'm not sure how to debug further.

**Describe the bug**
My avro data models are interacting very strangely with spark. When I create instances of the avro models an rdd.map function and serialize, all data is lost. Also, and maybe related, serializing one instance of a data class, causes another instance to serialize improperly. Observe:

**To Reproduce**

```
import re
from typing import Union
from dataclasses import dataclass

import typing

from dataclasses_avroschema import AvroModel


@dataclass
class Address(AvroModel):
    "An Address"
    street: str
    street_number: int



@dataclass
class User(AvroModel):
    "User with multiple Address"
    name: Union[str,int]
    age: int
    addresses: typing.List[Address]
address=Address.fake()
data_user = {
    "name": "john",
    "age": 20,
    "addresses": [address],
}

# create an User instance
user = User(**data_user)



def encode(x):
    return User(name='test',age=20,addresses=[Address(**i) for i in x['addresses']]).asdict()

df = [{'addresses':[Address.fake().asdict(),Address.fake().asdict()] } for _ in range(100)]

print('Try encoding without parallelizing: works just fine.',encode(df[0]))
df = spark.sparkContext.parallelize(df)
print('Parallelizing, collecting and then encoding: also works.',encode(df.first()))
print('Parallelizing, encoding, then collecting: all data is lost. ',df.map(lambda x: encode(x)).first())
print('Before using Spark user object serializes with no problems: ',user.serialize())
spark.sparkContext.parallelize([User.fake() for _ in range(100)]).first()
print('After parallelizing another object of the User class, the first user object no longer is able to serialize: ', user.serialize())`



```
which outputs:

```
Try encoding without parallelizing: works just fine. {'name': 'test', 'age': 20, 'addresses': [{'street': 'hwUGHxQIBVZJWDYbnTgo', 'street_number': 8731}, {'street': 'kCaMDTvXUsjbcpeMnhpB', 'street_number': 1857}]}
Parallelizing, collecting and then encoding: also works. {'name': 'test', 'age': 20, 'addresses': [{'street': 'hwUGHxQIBVZJWDYbnTgo', 'street_number': 8731}, {'street': 'kCaMDTvXUsjbcpeMnhpB', 'street_number': 1857}]}
Parallelizing, encoding, then collecting: all data is lost.  {}
Before using Spark user object serializes with no problems:  b'\x00\x08john(\x02(lirWCvQBSOsWswmvWFji\xca\x97\x01\x00'
After parallelizing another object of the User class, the first user object no longer is able to serialize:  b''
```

**Expected behavior**
To construct and serialize avro models from within a spark rdd/dataframe without losing data, especially when they have lists of other avro data models as part of their schema (though I am not 100% sure whether this is part of the problem).

**Comment by nathan-oplus** [comment]
Took me a few tries to name this bug. :sweat_smile:

**Comment by nathan-oplus** [comment]
Update: even if the "address" field has a list of strings and not a list of Addresses the same error occurs
```
import re
from typing import Union
from dataclasses import dataclass

import typing

from dataclasses_avroschema import AvroModel


@dataclass
class Address(AvroModel):
    "An Address"
    street: str
    street_number: int



@dataclass
class User(AvroModel):
    "User with multiple Address"
    name: Union[str,int]
    age: int
    addresses: typing.List[str]
address=Address.fake()
data_user = {
    "name": "john",
    "age": 20,
    "addresses": ['test'],
}

# create an User instance
user = User(**data_user)


# >>> b"\x08john(\x02\x08test\x14\x00"
def encode(x):
    return User(name='test',age=20,addresses=['test' for i in range(10)]).asdict()

df = [{'addresses':['test','test'] } for _ in range(100)]

print('Try encoding without parallelizing: works just fine.',encode(df[0]))
df = spark.sparkContext.parallelize(df)
print('Parallelizing, collecting and then encoding: also works.',encode(df.first()))
print('Parallelizing, encoding, then collecting: all data is lost. ',df.map(lambda x: encode(x)).first())
df = df.map(lambda x: {
    'user':x
    })
print('Before using Spark user object serializes with no problems: ',user.serialize())
spark.sparkContext.parallelize([User.fake() for _ in range(100)]).first()
print('After parallelizing another object of the User class, the first user object no longer is able to serialize: ', user.serialize())
```
```
Try encoding without parallelizing: works just fine. {'name': 'test', 'age': 20, 'addresses': ['test', 'test', 'test', 'test', 'test', 'test', 'test', 'test', 'test', 'test']}
Parallelizing, collecting and then encoding: also works. {'name': 'test', 'age': 20, 'addresses': ['test', 'test', 'test', 'test', 'test', 'test', 'test', 'test', 'test', 'test']}
Parallelizing, encoding, then collecting: all data is lost.  {}
Before using Spark user object serializes with no problems:  b'\x00\x08john(\x02\x08test\x00'
After parallelizing another object of the User class, the first user object no longer is able to serialize:  b''
```

**Comment by nathan-oplus** [comment]
Actually, wait, here I reproduce the bug with no attributes that are lists. 
 #```import re
from typing import Union
from dataclasses import dataclass

import typing

from dataclasses_avroschema import AvroModel




@dataclass
class User(AvroModel):
    "User with multiple Address"
    name: Union[str,int]
    age: int
    addresses: str

data_user = {
    "name": "john",
    "age": 20,
    "addresses": 'test',
}

# create an User instance
user = User(**data_user)


# >>> b"\x08john(\x02\x08test\x14\x00"
def encode(x):
    return User(name='test',age=20,addresses='test').asdict()

df = [{'addresses':['test','test'] } for _ in range(100)]

print('Try encoding without parallelizing: works just fine.',encode(df[0]))
df = spark.sparkContext.parallelize(df)
print('Parallelizing, collecting and then encoding: also works.',encode(df.first()))
print('Parallelizing, encoding, then collecting: all data is lost. ',df.map(lambda x: encode(x)).first())
print('Before using Spark user object serializes with no problems: ',user.serialize())
spark.sparkContext.parallelize([User.fake() for _ in range(100)]).first()
print('After parallelizing another object of the User class, the first user object no longer is able to serialize: ', user.serialize())```
```
Try encoding without parallelizing: works just fine. {'name': 'test', 'age': 20, 'addresses': 'test'}
Parallelizing, collecting and then encoding: also works. {'name': 'test', 'age': 20, 'addresses': 'test'}
Parallelizing, encoding, then collecting: all data is lost.  {}
Before using Spark user object serializes with no problems:  b'\x00\x08john(\x08test'
After parallelizing another object of the User class, the first user object no longer is able to serialize:  b''```

**Comment by marcosschroh** [comment]
Hi @nathan-oplus 

Thanks for reporting this problem. It looks like it is a "spark problem" regarding to `parallelizing`. I will try to take a look asap but I have never use `spark` before, so It might take a while

**Comment by nathan-oplus** [comment]
I agree about it being a "spark problem". Let me know if I can do anything to help.

**Comment by marcosschroh** [comment]
Hi @nathan-oplus 

After digging a bit with this problem I found that the same happens also for normal `dataclasses`.  I have the following script with a simple `encode` function that returns an user instance. The result is the expected one but if you try to return the dict representation of the user instance it always returns an empty dict `{}` (same result as the open issue). The reason for that strange behavior is that `parallelize` splits the load and some references are not pointing to the same object.

Inside the python `dataclasses.py` module the [fields function](https://github.com/python/cpython/blob/3.10/Lib/dataclasses.py#L1202) returns all the instance fields to be represented as a dictionary and when you run this function with `parallelize`  the `if f._field_type is _FIELD` always returns `False` (it does not find any fields). Why? Because the `The is keyword "is" used to test if two variables refer to the same object` which they are NOT.  I added a `print` statement inside the `python module` to check this (see the script output)

```python
from typing import Union

from pyspark.sql import SparkSession

import dataclasses


spark = SparkSession.builder.getOrCreate()


@dataclasses.dataclass
class User:
   "User with multiple Address"
   name: Union[str,int]
   age: int
   addresses: str

data_user = {
   "name": "john",
   "age": 20,
   "addresses": 'test',
}

# create an User instance
user = User(**data_user)


def encode(_):
   user = User(name="test", age=20, addresses="test")
   print(f"This is the user {user}. To dict: {dataclasses.asdict(user)}")
   return user


def main():
   df = [User(name="marcos", age=20, addresses="test") for _ in range(2)]
   df = spark.sparkContext.parallelize(df)

   print(df.map(lambda x: encode(x)).collect())
    
if __name__ == "__main__":
    main()
```

Output script:

```bash
❯ python3.10 example.py
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
23/04/26 18:19:26 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable

Field name name. Field _field_type reference 4454563072, dataclasses._FIELD reference 4454722160  # Point to different objects

Field name age. Field _field_type reference 4454563072, dataclasses._FIELD reference 4454722160 # Point to different objects

Field name addresses. Field _field_type reference 4454563072, dataclasses._FIELD reference 4454722160 # Point to different objects

This is the user User(name='test', age=20, addresses='test'). To dict: {}

Field name name. Field _field_type reference 4454563072, dataclasses._FIELD reference 4454722160  # Point to different objects

Field name age. Field _field_type reference 4454563072, dataclasses._FIELD reference 4454722160  # Point to different objects

Field name addresses. Field _field_type reference 4454563072, dataclasses._FIELD reference 4454722160  # Point to different objects

This is the user User(name='test', age=20, addresses='test'). To dict: {}
[User(name='test', age=20, addresses='test'), User(name='test', age=20, addresses='test')]
```

In summary, the `parallelize` function should not be used when we need to check if variables are pointing to the same address location. To solve your problem, just return the `User` instance rather than the `dict` representation.
